csrc/ir/nodes.cpp

```diff
@@ -4380,6 +4391,12 @@ std::vector<PolymorphicValue> SdpaFwdOp::evaluate(
               /*return_debug_mask=*/false,
               scale);
 
+  // Add back the device dim axis for outputs with a head dimension.
+  if (handle_device_dim) {
+    output = output.unsqueeze(0);
+    log_sumexp = log_sumexp.unsqueeze(0);
+  }
+
```

Comment:
This should be after L4401-4402 since dim=3 is assumed to be last dimension. Or `output.sizes()[3]` should be changed to `output.sizes()[-1]`